### DISCUSSION TITLE
Distributed multi-agent systems

### DISCUSSION THREAD

**Discussion by halllo** [microsoft/semantic-kernel#8323] [Q&A]
Is SK intended to support distributed multi-agent systems where the agents don’t run in-process but on different machines?

**Comment by evchaki** [comment]
Yes - each assistant is managed by its own processes.  If you use OpenAI / Azure OpenAI Assistant API you can see this in action.  Check out our Agent Framework - https://devblogs.microsoft.com/semantic-kernel/introducing-agents-in-semantic-kernel/

**Reply by halllo** [reply]
Cool, thanks a lot.

My thoughts are that if we have `AgentGroupChat` or `ChatHistory` centralised and thread-safe, we can allow multiple agents to work on them. To enable distributed agents / event-based agents, we would need some form of "remote" `AgentGroupChat` or `ChatHistory`.

What is the recommended best practice for having a centralised remote `AgentGroupChat` to which multiple processes can connect? Does it work over HTTP? Can we derive form it?

**Reply by evchaki** [reply]
You are correct, Open AI assistants have a Thread where multiple agents can see a shared chat history to work off of.  We showed this off during Office Hours this week.  The video should be out soon on our YouTube feed.  You can check out a .NET sample here - https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/GettingStartedWithAgents/Step4_KernelFunctionStrategies.cs

**Reply by halllo** [reply]
Very cool. I am looking forward watching that.

Do you plan to enable distributed agents / event-based agents with `ChatCompletionAgent` without OpenAI APIs? I am thinking about an interface that would allow me to "bring-my-own centralised threads API".

**Reply by halllo** [reply]
If `ChatHistory` was an interface we could use an implementation of a conflict-free replicated data type (CRDT). That might be an even better fit than a centralised threads API.
